src/factory.rs

```diff
@@ -1,26 +1,108 @@
+
     unsafe {
-        let mut code =
-            runtime::RoGetActivationFactory(HString::from(C::NAME).abi(), &I::iid(), &mut ptr);
+        // First attempt to get the activation factory via the OS.
+        let mut code = RoGetActivationFactory(name.abi(), &I::iid(), &mut factory);
 
+        // If this fails because combase hasn't been loaded yet then load combase
+        // automatically so that it "just works" for apartment-agnostic code.
         if code == ErrorCode::NOT_INITIALIZED {
             let mut _cookie = std::ptr::null_mut();
-            runtime::CoIncrementMTAUsage(&mut _cookie);
+            CoIncrementMTAUsage(&mut _cookie);
+
+            // Now try a second time to get the activation factory via the OS.
+            code = RoGetActivationFactory(name.abi(), &I::iid(), &mut factory);
+        }
 
-            code =
-                runtime::RoGetActivationFactory(HString::from(C::NAME).abi(), &I::iid(), &mut ptr);
+        // If this succeeded then retun the resulting factory interface.
+        if code.is_ok() {
+            return Ok(std::mem::transmute_copy(&factory));
         }
 
-        code.and_then(|| std::mem::transmute_copy(&ptr))
+        // If not, first capture the error information from the failure above so that we
+        // can ultimately return this error information if all else fails.
+        let original: Error = code.into();
+
+        // Now attempt to find the factory's implementation heuristically.
+        let mut path = C::NAME;
+
+        // Remove the suffix until a match is found. For example, if the class name is
+        // "A.B.TypeName" then this will first attempt to load "A.B.dll" and if that
+        // fails it will attempt "A.dll" before giving up.
+        while let Some(pos) = path.rfind('.') {
+            path = &path[..pos];
+
+            // Turn the resulting namespace portion into a DLL name.
+            let path: Vec<u16> = path
+                .encode_utf16()
+                .chain(".dll".encode_utf16())
+                .chain(std::iter::once(0))
+                .collect();
+
+            // Attempt to load the DLL.
+            let library: Library = LoadLibraryW(path.as_ptr()).into();
+
+            if library.handle.is_null() {
+                continue;
+            }
+
+            // If the DLL was found then get the export used to retrieve the factory.
+            let library_call =
+                GetProcAddress(library.handle, b"DllGetActivationFactory\0".as_ptr());
+
+            if library_call.is_null() {
+                continue;
+            }
+
+            let library_call: DllGetActivationFactory = std::mem::transmute(library_call);
+            let mut default_factory: IActivationFactory = std::mem::zeroed();
+
+            // Now call DllGetActivationFactory to request the given class.
+            if library_call(name.abi(), default_factory.set_abi()).is_err() {
+                continue;
+            }
```

Comment:
Might be paranoia, but should we `debug_assert` that default_factory is not null?